text/0000-metric-label-set.md

```diff
@@ -0,0 +1,123 @@
+	      (Cumulative, Value),
+	      (Measure, Value),
+	      ... ])
+```
+
+This RFC proposes the new `LabelSet` concept is returned by an API named `Meter.DefineLabels({ Key: Value, ... })` which allows the SDK to potentially sort, canonicalize, or hash the set of labels once, allowing it to be re-used in several ways.  For example, the `RecordBatch` above can be written as one call to `DefineLabels` and inidividual operations.
+
+```
+Labels := meter.DefineLabels({ Key1: Value1,
+       	  		       Key2: Value2,
+			       ... })
+Cumulative.Add(Value, Labels)
+Gauge.Set(Value, Labels)
+Measure.Record(Value, Labels)
+...
+```
+
+With a first-class `LabelSet`, labels can even be re-used across multiple calls to `RecordBatch`.
+
+## Explanation
+
+Metric instrument APIs which presently take labels in the form `{ Key: Value, ... }` will be updated to take an explicit `LabelSet`.  The `Meter.DefineLabels` API method supports getting a `LabelSet` from the SDK, allowing the programmer to pre-define labels without being required to manage handles.  This brings the number of ways to update a metric to three, via re-using a handle from `GetHandle()`:
+
+```
+cumulative := metric.NewFloat64Cumulative("my_counter", [
+					    "required_key1",
+					    "required_key2",
+					  ])
+labels := meter.DefineLabels({ "required_key1": value1,
+			       "required_key2": value2 })
+handle := cumulative.GetHandle(labels)
+for ... {
+   handle.Add(quantity)
+}
+```
+
+To operate on a metric instrument directly, without requiring a handle:
+
+```
+cumulative.Add(quantity, labels)
+gauge.Set(quantity, labels)
+measure.Record(quantity, labels)
+```
+
+To operate on a batch of labels,
+
+```
+RecordBatch(labels, [
+	(Instrument1, Value1),
+	(Instrument2, Value2),
+	...
+      ])
+```
+
+Note that `meter.GetHandle(instrument, meter.DefineLabels())` is the same as `meter.GetDefaultHandle()`.
+
+### Ordered `LabelSet` option
+
+OpenCensus specified support for ordered label sets, in which values for required keys may be passed in order, allowing the implementation directly compute a unique encoding for the label set.   We recognize that providing ordered values is an option that makes sense in some languages and not others, and leave this as an option.  Support for ordered label sets might be arranged through ordered key sets, for example:
+
+```
+requiredKeys := metric.DefineKeys("key1", "key2", "key3")
+labelSet := requiredKeys.DefineValues(meter, 1, 2, 3)
+```
+
+## Internal details
+
+Metric instruments are specified as SDK-independent objects, therefore metric handles were required to bind the instrument to the SDK in order to operate. In this proposal, `LabelSet` becomes responsible for binding the SDK at any call site where it is used.  Other than knowing the `Meter` that defined it, `LabelSet` is an opaque interface.  The three ways to use `LabelSet` in the metrics API are:
+
+```
+instrument.GetHandle(labels).Action(value)       // Action on a handle 
```

Comment:
are there still advantage for exposing a handles concept? I think as a user, passing in a labelset to a metric and having it resolve to a handle under the hood is a lot simpler than explicitly instantiating and incrementing handle objects.

Reply:
This has been emphasized as a requirement to achieve prometheus-level performance in critical sections of code (e.g., gRPC). I wrote "as fast as a few machine instructions" to capture this requirement.

I will add some discussion about the difference in cost of not using handles. I believe the "under the hood" computation is just a map-lookup where the key is a pair of (instrument ID, `LabelSet` ID). 

Some SDKs will prefer not to support this, but it's trivial to supply an implementation of these handles that call through to `RecordBatch(LabelSet, Measurement{Instrument, Value})` in ~50 lines of code.